source/adapters/hip/image.cpp

```diff
@@ -8,178 +8,1335 @@
+        cpy_desc.Height =
+            std::max(MinCopyHeight, pCopyRegion->copyExtent.height);
+        cpy_desc.Depth = pCopyRegion->copyExtent.depth;
+        // 'hipMemcpy3DAsync' requires us to correctly create 'hipMemcpy3DParms'
+        // struct object which adds a little complexity (e.g. 'hipPitchedPtr').
+        UR_CHECK_ERROR(hipDrvMemcpy3DAsync(&cpy_desc, Stream));
+      }
+    } else if (imageCopyFlags == UR_EXP_IMAGE_COPY_FLAG_DEVICE_TO_HOST) {
+      if (pSrcImageDesc->type == UR_MEM_TYPE_IMAGE1D) {
+        unsigned int memType{};
+        // hipPointerGetAttribute can detect HIP-registered memory of types:
+        // hipMemoryTypeHost, hipMemoryTypeDevice, or hipMemoryTypeArray.
+        UR_CHECK_ERROR(
+            hipPointerGetAttribute(&memType, HIP_POINTER_ATTRIBUTE_MEMORY_TYPE,
+                                   reinterpret_cast<hipDeviceptr_t>(pDst)));
+        UR_ASSERT(memType == hipMemoryTypeDevice ||
+                      memType == hipMemoryTypeArray,
+                  UR_RESULT_ERROR_INVALID_VALUE);
+
+        size_t CopyExtentBytes = PixelSizeBytes * pCopyRegion->copyExtent.width;
+        void *DstWithOffset =
+            static_cast<void *>(static_cast<char *>(pDst) +
+                                (PixelSizeBytes * pCopyRegion->dstOffset.x));
+
+        if (memType == hipMemoryTypeArray) {
+          // HIP doesn not provide async copies between image arrays and host
+          // memory in versions early than 6.2.
+#if HIP_VERSION >= 60200000
+          UR_CHECK_ERROR(hipMemcpyAtoHAsync(
+              DstWithOffset, static_cast<hipArray_t>(const_cast<void *>(pSrc)),
+              PixelSizeBytes * pCopyRegion->srcOffset.x, CopyExtentBytes,
+              Stream));
+#else
+          UR_CHECK_ERROR(hipMemcpyAtoH(
+              DstWithOffset, static_cast<hipArray_t>(const_cast<void *>(pSrc)),
+              PixelSizeBytes * pCopyRegion->srcOffset.x, CopyExtentBytes));
+#endif
+        } else if (memType == hipMemoryTypeDevice) {
+          const char *SrcWithOffset =
+              static_cast<const char *>(pSrc) +
+              (pCopyRegion->srcOffset.x * PixelSizeBytes);
+          UR_CHECK_ERROR(hipMemcpyDtoHAsync(
+              DstWithOffset,
+              static_cast<hipDeviceptr_t>(const_cast<char *>(SrcWithOffset)),
+              CopyExtentBytes, Stream));
+        } else {
+          // This should be unreachable.
+          return UR_RESULT_ERROR_INVALID_VALUE;
+        }
+      } else if (pSrcImageDesc->type == UR_MEM_TYPE_IMAGE2D) {
+        hip_Memcpy2D cpy_desc = {};
+        cpy_desc.srcXInBytes = pCopyRegion->srcOffset.x * PixelSizeBytes;
+        cpy_desc.srcY = pCopyRegion->srcOffset.y;
+        cpy_desc.dstXInBytes = pCopyRegion->dstOffset.x * PixelSizeBytes;
+        cpy_desc.dstY = pCopyRegion->dstOffset.y;
+        cpy_desc.dstMemoryType = hipMemoryTypeHost;
+        cpy_desc.dstHost = pDst;
+        if (pSrcImageDesc->rowPitch == 0) {
```

Comment:
Is it okay that `cpy_desc.srcPitch` is uninitialized in this path? I might have missed similar instances of this question before/after.

Reply:
So I think here we want it (`srcPitch`) zero-initialised. I thought it would be init to that, originally, from the looking at the `hip_Memcpy2D cpy_desc = {};` construction (another one taken from the Cuda adapter counterpart). That was on the idea that, assuming that's a POD type, it's all zero-init. However, if my understanding is wrong it may be better to do it explictly.

Reply:
In cuda `CU_MEMORYTYPE_ARRAY`, which is equivalent to `hipMemoryTypeArray`, arrays are opaque memory and do not have (at least exposed) notion of pitch.

So it is not set as it is not expected for the pitch field to be used at all.

Reply:
Yeah that's correct. However, I think Fraser's concern was a little more general here - that we might end up with a garbage value if left unitialised (which isn't good practice) - ideally we want it zero-init in such cases. Luckily, since C++11 with the brace-init syntax via the uniform init-brackets `{}` that are used for `cpy_desc` here (considering `hip_Memcpy2D` is a POD or trivially default-constructible), the members that are not explicitly initialised should get zero-init (or `nullptr` for pointers).